Add dependency description format document

mmitche opened the pull request:
This has been dealt with in email already, but would appreciate more feedback.

mmitche commented:
@davidsh fyi

chcosta commented on Documentation/DependencyDescriptionFormat.md:
```diff
@@ -0,0 +1,159 @@
+- **1 Dependency Details File** - XML file containing a set of information for each input dependency:
+  - Dependency Name – Unique, typically the package name, but it is not required that individual packages be listed.  For example, a repo which ingests 10 assets with the same version could group all under a single dependency.
+  -Version – Version of dependency
+  - URL – URL of repository (typically) that dependency the asset.
```
awkward "that dependency the asset"

jcagme commented on Documentation/DependencyDescriptionFormat.md:
```diff
@@ -0,0 +1,159 @@
+The dependency description is comprised of two types of assets:
+- **1 Dependency Details File** - XML file containing a set of information for each input dependency:
+  - Dependency Name – Unique, typically the package name, but it is not required that individual packages be listed.  For example, a repo which ingests 10 assets with the same version could group all under a single dependency.
+  -Version – Version of dependency
```
Leave space between - and V so it is a bullet point

jcagme commented on Documentation/DependencyDescriptionFormat.md:
```diff
@@ -0,0 +1,159 @@
+
+  The expression file corresponding to a dependency is implied by the dependency name and type.
+  - Product dependencies are always expressed in the version props file.
+  - Build dependencies either have specially handled names, which implies that other files expression the version, or they have an non-special name name, which implies that the version props file expresses the version. The following special names are in use:
```
name name

chcosta commented:
Have you heard any implementation details for how ["repo tools"](https://github.com/dotnet/cli/issues/8934#issuecomment-377271862) are going to work?  I'd imagine that would be a feature we may be interested in when it's available.

chcosta commented on Documentation/DependencyDescriptionFormat.md:
```diff
@@ -0,0 +1,159 @@
+  -Version – Version of dependency
+  - URL – URL of repository (typically) that dependency the asset.
+  - SHA – Git SHA at which the dependency was produced
+  - Dependency Class – Build or Product
```
What's the definition of a "Dependency Class" and why is it necessary to differentiate between Build and Product?

chcosta commented on Documentation/DependencyDescriptionFormat.md:
```diff
@@ -0,0 +1,159 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Dependencies>
+    <!-- Elements contains all product dependencies -->
+    <ProductDependencies>
```
I don't know what other classes of dependencies there would be (besides "Product" and "Build"), but it may make this model more extensible if the "Class" was an element of the "Dependency".

```
<Dependencies>
  <Dependency Name="dependencyA" Version="1.2.3-45" Class="Product">
    <SourceUri>https://github.com/dotnet/arepo</SourceUri> 
    <SourceSha>23498123740982349182340981234</SourceSha> 
  </Dependency> 
  <Dependency Name="DependencyB" Version="1.2.3-45" Class="Product"> 
     <SourceUri>https://github.com/dotnet/arepo</SourceUri> 
     <SourceSha>13242134123412341465</SourceSha> 
   </Dependency>
   ...
   <Dependency Name="Arcade.Sdk" Version="1.0.0" Class="Build"> 
      <SourceUri>https://github.com/dotnet/arcade</SourceUri> 
      <SourceSha>132412342341234234</SourceSha> 
   </Dependency>
</Dependencies>
```

mmitche replied:
I think the dependency format is a little more readable with higher level classes.  It encourages separation of build and product dependencies.  I'm also not sure what other dependency type there might be....Test maybe?

weshaggard commented on Documentation/DependencyDescriptionFormat.md:
```diff
@@ -0,0 +1,158 @@
+  - Version – Version of dependency
+  - URL – URL of repository (typically) that produced the asset.
+  - SHA – Git SHA at which the dependency was produced
+  - Dependency Class – Either 'Build' or 'Product'.  'Build' dependencies are those that are used to produce the product, 'Product' dependencies are effectively everything else.  It is useful to differentiate between these so that we know how we are building (e.g. what CLI SDKs are in use across all repositories).  Defined another way, two successive builds of a Build dependency on the same sha c ...
```
'Build' is a little ambiguous I wonder if you should call it "tools" or "toolset" instead?

weshaggard commented on Documentation/DependencyDescriptionFormat.md:
````diff
@@ -0,0 +1,158 @@
+```
+
+### Expression Files
+**Version.props (eng\Version.props)**
````
Will there be an expression file for the native bootstrapping?

mmitche replied:
Likely yes.

mmitche commented:
Updated:
- 'Build' is now 'Toolset'
- Added native toolset dependency expression file that @chcosta is working on.
- Clarify that additional dependency classes could be added if needed.  (e.g. test dependencies)

JohnTortugo commented on Documentation/DependencyDescriptionFormat.md:
```diff
@@ -0,0 +1,159 @@
+  - Version – Version of dependency
+  - URL – URL of repository (typically) that produced the asset.
+  - SHA – Git SHA at which the dependency was produced
+  - Dependency Class – Either 'Toolset' or 'Product'.  'Toolset' dependencies are those that are used to produce the product, 'Product' dependencies are effectively everything else.  It is useful to differentiate between these so that we know how we are building (e.g. what CLI SDKs are in use across all repositories).  Defined another way, two successive builds of a Toolset dependency on the same ...
```
Typo: thse

JohnTortugo commented on Documentation/DependencyDescriptionFormat.md:
```diff
@@ -0,0 +1,159 @@
+    <ProductDependencies>
+        <-- All product dependencies are contained in Version.Props -->
+        <Dependency Name="DependencyA" Version="1.2.3-45">
+            <SourceUri>https://github.com/dotnet/arepo</SourceUri>
```
I don't know if these are just examples or also a proposal for XML Schema. If all child elements of <Dependency> start with <Source*...> maybe we could factor out that prefix.

JohnTortugo commented on Documentation/DependencyDescriptionFormat.md:
```diff
@@ -0,0 +1,159 @@
+  <PropertyGroup>
+    <!-- DependencyA, DependencyB, DependencyC substrings correspond to
+         DependencyName elements in Version.Details.xml file -->
+    <DependencyAPackageVersion>4.5.0-preview2-26403-05</DependencyAPackageVersion>
```
I had understood that the version of the asset that a tool depends on would be stored completely in Version.Details.xml -> Dependency -> Version ; So, I am puzzled what does the **content** of the &lt;DependencyAPackageVersion%gt; means here.

That is, I had understood that anywhere I referenced &lt;DependencyA/&gt; I would be depending on asset DependencyA version 1.2.3-45. What did I miss?

mmitche replied:
Version.Details.xml cannot be ingested by all tools, especially msbuild.  So we have to use other expressions of those details as input to required tools.  So for DependencyA, your project file would do:

```<PackageReference Include="DependencyA" Version="$(DependencyAPackageVersion)" />```
